src/types/widget.ts

```diff
@@ -300,7 +284,73 @@ export type Widget = {
     state: SearchParameters,
     widgetSearchParametersOptions: WidgetSearchParametersOptions
   ): SearchParameters;
-};
+} & (
```

Comment:
Generics were created to solve this specific problem. They're a solution to define an interface that we implement in our widgets, and that users can implement in their own widgets. Here we loose this API definition.

Why avoiding them?

Reply:
I'm not avoiding them, but was trying to compare them and I literally didn't know what's better and what has which drawbacks. I think https://github.com/algolia/instantsearch.js/pull/4518#discussion_r498073495 makes it enough to choose generics. Let me know if you have more thoughts on the drawbacks of this approach (discriminated type union)  in addition to that.

Reply:
@francoischalifour I'm not sure what API definition we are loosing here. Also Generic have their own limitations and in our case we want to have stricter definition on our own widget than what we enforce for custom ones. This is where Generics find their limits here, forcing us to be less strict on our own widgets. I find discriminated type unions here to fit perfectly for what we want to accomplish while keeping the types more readable. For example here's what I have in mind: https://github.com/algolia/instantsearch.js/blob/311654f856a4b3625395a370e95ca606806efb6d/src/types/widget.ts#L237-L475

Reply:
I think a generic for RenderState defaulting to any / an object / similar would make sense though, that way external users can still use very loose typescript if they want, but can also opt-in to the same safety as us. We won't be able to put custom widgets in uiState & render state without a separate generic on InstantSearch I guess

Reply:
@yannickcr Can you help me understand, without generics, but with discriminated type unions, how to enable customers to create custom widgets and still let them implement `getWidgetRenderState` without type issue?

Reply:
@eunjae-lee With discriminated type unions a custom widget created by the user will not have a `$$type` property, so it will match the `GenericWidget` type.

In that case if the user wants to implement `getWidgetRenderState` he will need to follow the type of `getWidgetRenderState` as implemented in `GenericWidget` (`getWidgetRenderState?: (renderOptions: InitOptions | RenderOptions) => any;` here but as @Haroenv we could force it to return an object)

Reply:
Oh I didn't see you used `interface` to extend the `GenericWidget`.